File: bench.rb
def image_ramp_green(n)
  img = []
  f = 255.0 / (n - 1)
  (0...n).each do |i|
    img << {
      red: 0,
      green: (i * f).floor,
      blue: 0,
      alpha: 255
    }
  end
  img
end

def image_to_gray(img, n)
  (0...n).each do |i|
    y = (0.3 * img[i][:red] + 0.59 * img[i][:green] + 0.11 * img[i][:blue]).floor
    img[i][:red] = y
    img[i][:green] = y
    img[i][:blue] = y
  end
end

N = 400 * 400
img = image_ramp_green(N)

1000.times do
  image_to_gray(img, N)
end

File: bench.go
package main

import (
	"math"
)

type Pixel struct {
	Red   uint8
	Green uint8
	Blue  uint8
	Alpha uint8
}

func imageRampGreen(n int) []Pixel {
	img := make([]Pixel, n)
	f := 255.0 / float64(n-1)
	for i := 0; i < n; i++ {
		img[i] = Pixel{
			Red:   0,
			Green: uint8(math.Floor(float64(i) * f)),
			Blue:  0,
			Alpha: 255,
		}
	}
	return img
}

func imageToGray(img []Pixel, n int) {
	for i := 0; i < n; i++ {
		y := uint8(math.Floor(0.3*float64(img[i].Red) + 0.59*float64(img[i].Green) + 0.11*float64(img[i].Blue)))
		img[i].Red = y
		img[i].Green = y
		img[i].Blue = y
	}
}

func main() {
	N := 400 * 400
	img := imageRampGreen(N)
	for j := 0; j < 1000; j++ {
		imageToGray(img, N)
	}
}
